libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1074 @@
+                                                       G_DBUS_CALL_FLAGS_NONE,
+                                                       -1,
+                                                       cancellable,
+                                                       error);
+  g_autoptr(GVariant) shards_variant = NULL;
+  g_autoptr(GVariant) models_variant = NULL;
+  g_auto(GStrv) shards_strv = NULL;
+  g_autoptr(GPtrArray) model_hts = NULL;
+  GVariantIter iter;
+  GVariant *model_variant = NULL;
+
+  if (result == NULL)
+    return NULL;
+
+  g_variant_get (result, "(^as@aa{ss})", &shards_strv, &models_variant);
+
+  model_hts = g_ptr_array_new_full (g_variant_n_children (models_variant),
+                                    (GDestroyNotify) g_hash_table_unref);
+
+  g_variant_iter_init (&iter, models_variant);
+  while (g_variant_iter_loop (&iter, "@a{ss}", &model_variant))
+    g_ptr_array_add (model_hts,
+                     hashtable_from_variant (model_variant));
+
+  /* Now that we have the models and shards, we can marshal them into
+   * a GPtrArray containing the discovery-feed models */
+  model_data_array = marshal_func ((const gchar * const *) shards_strv,
+                                   model_hts,
+                                   marshal_data);
+
+  return g_steal_pointer (&model_data_array);
+}
+
+static GObject *
+call_dbus_proxy_and_construct_from_model (GDBusProxy           *proxy,
+                                          const gchar          *method_name,
+                                          const gchar          *source_name,
+                                          ModelFromResultFunc   marshal_func,
+                                          gpointer              marshal_data,
+                                          GCancellable         *cancellable,
+                                          GError              **error)
+{
+  g_autoptr(GPtrArray) model_data_array = NULL;
+  g_autoptr(GVariant) result = g_dbus_proxy_call_sync (proxy,
+                                                       method_name,
+                                                       NULL,
+                                                       G_DBUS_CALL_FLAGS_NONE,
+                                                       -1,
+                                                       cancellable,
+                                                       error);
+  g_autoptr(GVariant) model_variant = NULL;
+  g_autoptr(GHashTable) model_ht = NULL;
+
+  if (result == NULL)
+    return NULL;
+
+  g_variant_get (result, "(@a{ss})", &model_variant);
+  model_ht = hashtable_from_variant (model_variant);
+
+  /* Now that we have the model, we can marshal it into a
+   * GObject */
+  return marshal_func (model_ht, marshal_data);
+}
+
+/* Quite surprising that I could not find an alternative to this in Gio */
+static gchar *
+remove_uri_prefix (const gchar  *uri,
+                   GError      **error)
```

Comment:
Hmm, what format are these URIs in? This seems to work:
```
gjs> var f = Gio.File.new_for_uri('ekn:///0xdeadbeef');
gjs> var p = Gio.File.new_for_uri('ekn:///');
gjs> p.get_relative_path(f);
"0xdeadbeef"
```

And this too
```
gjs> var f = Gio.File.new_for_uri('ekn:0xdeadbeef');
gjs> var p = Gio.File.new_for_uri('ekn:///');
gjs> p.get_relative_path(f);
"0xdeadbeef"
```

Reply:
Ah interesting, I didn't think that `get_relative_path` would work in that way. I'll use that then.

Reply:
Okay, strange. So it seems that `g_file_get_relative_path` works if I do

```
g_autoptr(GFile) parent = g_file_new_for_uri ("ekn://")
g_autoptr(GFile) child = g_file_new_for_uri (uri);

return g_file_get_relative_path (parent, child);
```

But not:

```
g_autoptr(GFile) parent = g_file_new_for_uri ("ekn:///")
g_autoptr(GFile) child = g_file_new_for_uri (uri);

return g_file_get_relative_path (parent, child);
```

However, it works just fine if I do the former from Gjs! Could be something to do with different gio version from within flatpak and on the system I suppose.

For now I've just used the former.